### DISCUSSION TITLE
Nxt editor to Motion Builder

### DISCUSSION THREAD

**Discussion by borparo** [nxt-dev/nxt_editor#249] [Q&A]
Hi!
I just discovered Nxt and I've been trying to use it with Motion Builder. So far, because can be installed trough pip I can use mobupy to download it and use it within the Python Editor that comes with Mobu (no in a venv though). Basically, import nxt and run the execute_graph method, will work.  I can create a graph with nodes containing pyfbsdk syntax and  because is inside Motion Builder it goes smoothly.

So, my next step was trying to use the editor in a venv, but I can't make a connection  with the  editor to mobu. As far as I know,  the way to  send python commands to mobu from an external program is connecting to mobu through telnet. 
So this code:
```
import telnetlib
host =  '127.0.0.1'
port = 4242
tn = telnetlib.Telnet(host, port)
tn.write('cube=FBModelCube("box")'.encode('ascii') + b'\n')
tn.write('cube.Show = True'.encode('ascii') + b'\n')
tn.write('cube.Scaling = FBVector3d(30, 30, 30)'.encode('ascii') + b'\n')
```
run from the IDLE will create a cube in mobu. In the Nxt editor I created a node with the same code. Executing Selected  doesn't report any error, but when switching to Mobu, nothing happened in the scene.

Hope someone can point me in the right direction.
Thanks !
Borja

**Comment by borparo** [comment]
by the way,  when trying to run `nxt ui` from the location mobupy installed it I got the following errors:
```
PS C:\Users\Borja\AppData\Roaming\Python\Python37\Scripts> ./nxt ui
Traceback (most recent call last):
  File "c:\program files\autodesk\motionbuilder 2022\bin\x64\python37\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "c:\program files\autodesk\motionbuilder 2022\bin\x64\python37\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\Borja\AppData\Roaming\Python\Python37\Scripts\pyside2-rcc.exe\__main__.py", line 4, in <module>
ModuleNotFoundError: No module named 'PySide2.scripts'
Traceback (most recent call last):
  File "C:\Users\Borja\AppData\Roaming\Python\Python37\site-packages\nxt_editor\__init__.py", line 88, in <module>
    from nxt_editor import qresources
ImportError: cannot import name 'qresources' from 'nxt_editor' (C:\Users\Borja\AppData\Roaming\Python\Python37\site-packages\nxt_editor\__init__.py)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\Borja\AppData\Roaming\Python\Python37\site-packages\nxt_editor\__init__.py", line 77, in make_resources
    '-o', result_path], cwd=pyside_dir)
  File "C:\Program Files\Autodesk\MotionBuilder 2022\bin\x64\python37\lib\subprocess.py", line 358, in check_call
    retcode = call(*popenargs, **kwargs)
  File "C:\Program Files\Autodesk\MotionBuilder 2022\bin\x64\python37\lib\subprocess.py", line 339, in call
    with Popen(*popenargs, **kwargs) as p:
  File "C:\Program Files\Autodesk\MotionBuilder 2022\bin\x64\python37\lib\subprocess.py", line 800, in __init__
    restore_signals, start_new_session)
  File "C:\Program Files\Autodesk\MotionBuilder 2022\bin\x64\python37\lib\subprocess.py", line 1207, in _execute_child
    startupinfo)
FileNotFoundError: [WinError 2] The system cannot find the file specified

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\Borja\AppData\Roaming\Python\Python37\Scripts\nxt-script.py", line 11, in <module>
    load_entry_point('nxt-core==0.14.0', 'console_scripts', 'nxt')()
  File "C:\Program Files\Autodesk\MotionBuilder 2022\bin\x64\python37\lib\site-packages\pkg_resources\__init__.py", line 489, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "C:\Program Files\Autodesk\MotionBuilder 2022\bin\x64\python37\lib\site-packages\pkg_resources\__init__.py", line 2852, in load_entry_point
    return ep.load()
  File "C:\Program Files\Autodesk\MotionBuilder 2022\bin\x64\python37\lib\site-packages\pkg_resources\__init__.py", line 2443, in load
    return self.resolve()
  File "C:\Program Files\Autodesk\MotionBuilder 2022\bin\x64\python37\lib\site-packages\pkg_resources\__init__.py", line 2449, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "C:\Users\Borja\AppData\Roaming\Python\Python37\site-packages\nxt\cli.py", line 19, in <module>
    import nxt_editor
  File "C:\Users\Borja\AppData\Roaming\Python\Python37\site-packages\nxt_editor\__init__.py", line 90, in <module>
    make_resources()
  File "C:\Users\Borja\AppData\Roaming\Python\Python37\site-packages\nxt_editor\__init__.py", line 82, in make_resources
    "output to \"{}\"".format(qrc_path, result_path))
Exception: Failed to generate UI resources using pyside2 rcc! Reinstalling pyside2 may fix the problem. If you know how to use rcc please build from: "C:\Users\Borja\AppData\Roaming\Python\Python37\site-packages\nxt_editor\resources/resources.qrc" and output to "C:\Users\Borja\AppData\Roaming\Python\Python37\site-packages\nxt_editor\qresources.py"
PS C:\Users\Borja\AppData\Roaming\Python\Python37\Scripts> dir

```
I tried to use pyside2-rcc, with no luck:
```
PS C:\Users\Borja\AppData\Roaming\Python\Python37\Scripts> .\pyside2-rcc C:\Users\Borja\AppData\Roaming\Python\Python37\site-packages\nxt_editor\resources\resources.qrc -o C:\Users\Borja\AppData\Roaming\Python\Python37\site-packages\nxt_editor\qresources.py
Traceback (most recent call last):
  File "c:\program files\autodesk\motionbuilder 2022\bin\x64\python37\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "c:\program files\autodesk\motionbuilder 2022\bin\x64\python37\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\Borja\AppData\Roaming\Python\Python37\Scripts\pyside2-rcc.exe\__main__.py", line 4, in <module>
ModuleNotFoundError: No module named 'PySide2.scripts'
PS C:\Users\Borja\AppData\Roaming\Python\Python37\Scripts>

```
looking into the PySide2 directory, I can see the scripts folder:
![image](https://user-images.githubusercontent.com/7436314/180298846-27b06276-cce7-4625-a4f7-9be1ec71fa9d.png)

**Comment by ImLucasBrown** [comment]
Hi Borja, 
Sorry to hear you're having trouble getting the qresources built. I had a similar issue when working on the Blender integration.
I don't have access to Motion Builder right now, but here is where I would start:
- Build a new virtual environment using the same version of Python you're using in Motion Builder.
   - If you're using Python 2.x please follow the instructions [here](https://github.com/nxt-dev/nxt_editor/blob/release/CONTRIBUTING.md#python-environment) for installing, using Conda will allow you to get a Py2 version of Pyside2
- `pip install nxt-editor`
- `nxt ui`
- Once it's opened once, navigate to the install location and copy out the `qrecourse.py` into your Motion Builder NXT install location.

That should work, assuming the standalone install of NXT is using the _same major version_ of Python.
_Because NXT runs on so many different platforms and permeations of Python environments, we don't vendor that resource file._

Good luck! Let me know if you're able to solve the issue.
-Lucas

**Reply by borparo** [reply]
Hi Lucas,
Apologies for the late answer, and thanks a lot for the help. I  tried your steps, and couldn't make it work.  The nxt_editor window shows with a white background and then closes and I get the following line in the shell.
```
PS C:\Users\Borja\AppData\Roaming\Python\Python37\Scripts> ./nxt ui
<nxt_editor.main_window.MainWindow(0x2a69ef05370, name="Main Window") at 0x000002A6A110E7C8>

PS C:\Users\Borja\AppData\Roaming\Python\Python39\Scripts> ./nxt ui
<nxt_editor.main_window.MainWindow(0x2ac4e8fe710, name="Main Window") at 0x000002AC4D910480>
```
I gave it a few tries in the past days, with a clean install of Python and Mobu (for versions 2022 and 2033) but the result is the same in both cases. I will try again through the weekend, and see what can find out :)
Kind regards,
Borja

**Comment by borparo** [comment]
Hello again Lucas:
I attached the log file of my last try. The editor so far freezes after starting the server. Then, after some minutes have to kill the process because unresposiveness.
[nxt_session_1659278662.log](https://github.com/nxt-dev/nxt_editor/files/9229089/nxt_session_1659278662.log)

Kind regards,
Borja

**Reply by ImLucasBrown** [reply]
Right, I totally forgot to mention, you need to pass the `-no-rpc` flag if you use the CLI. 
If you're starting the main window manually from a script or something make sure to pass `start_rpc=False` to the [main window init.](https://github.com/nxt-dev/nxt_editor/blob/a8b38df1e6fa649783e2cb5a92c524bb861f43f8/nxt_editor/main_window.py#L53)
Another way to prevent the RPC startup is to set the environment variable `NXT_DCC` to something like`MotionBuilder`, this will only work if you're using the `nxt.session` module to start it.
Long story short, the RPC code only reliably works in "vanilla" Python, not interpreters vendored with DCCs.
